#include <ros/ros.h>
#include <ros/package.h>
#include <body_msgs/Skeletons.h>
#include <gesture/HandPos.h>
#include <gesture/gesture.h>
#include <fstream>
#include "floatfann.h"
#define PI 3.14159
char str[] ="training.data";
int done=0;
void chatterCallback(const gesture::HandPos::ConstPtr& msg )
{
  static bool flag[2];
  static int i,j;
  static bool state;
  int parts1,extra1,k1,parts2,extra2,k2;
  int m1=0,m2=0;

  static float frame0[900][2],frame1[900][2];
  static float data1[9][2],data2[9][2];
  float a1=0,b1=0,c1=0,d1=0,e1=0,cos1=0,sin1=0,a2=0,b2=0,c2=0,d2=0,e2=0,cos2=0,sin2=0;
  fann_type f1[8],f2[8];

// zero indexed is left hand, one indexed is right hand

 std::ofstream outfile(str);

 if(!state){
	  if(msg->left_click||msg->right_click){
		
		if(msg->left_click&&!msg->right_click){
		flag[0]=1;
		  frame0[i][0]=msg->left_position.x;
		  frame0[i][1]=msg->left_position.y;
		  ROS_INFO("%d %f %f",i,frame0[i][0],frame0[i][1]);
		if(i+1!=900)
			i++;
		else{
			ROS_INFO("Buffer Full..Take lesser time to complete your gesture");		
			exit;
			}		
		}

		if(!msg->left_click&&msg->right_click){
		flag[1]=1;
		  frame1[j][0]=msg->right_position.x;
		  frame1[j][1]=msg->right_position.y;
		  ROS_INFO("%d %f %f",j,frame1[j][0],frame1[j][1]);
		if(j+1!=900)
			j++;
		else{
			ROS_INFO("Buffer Full..Take lesser time to complete your gesture");		
			exit;
			}
		}

		if(msg->left_click&&msg->right_click){
		flag[1]=1;
		flag[0]=1;
		  frame0[i][0]=msg->left_position.x;
		  frame0[i][1]=msg->left_position.y;
		  frame1[j][0]=msg->right_position.x;
		  frame1[j][1]=msg->right_position.y;
		  ROS_INFO("BOTH IN : %d %f %f %d %f %f",i,frame0[i][0],frame0[i][1],j,frame1[j][0],frame1[j][1]);
		if(i+1!=900)
			i++;
		else{
			ROS_INFO("Buffer Full..Take lesser time to complete your gesture");		
			exit;
			}
		if(j+1!=900)
			j++;
		else{
			ROS_INFO("Buffer Full..Take lesser time to complete your gesture");		
			exit;
			}
		}
		
	  }
	  if(!msg->left_click && !msg->right_click && !flag[0]&&!flag[1]) {
		  ROS_INFO("Enter the Zone %f || %f is less",msg->left_position.z,msg->right_position.z);
			i=0;
	  }
	  if(!msg->left_click && !msg->right_click && (flag[0]||flag[1])) {
	  		  ROS_INFO("Gesture Done");
	  		  state=1;
	  	  }
  }
 else {
	if(flag[0]&&!flag[1]){	 //pick for left hand
	   parts1=(i-8)/10;
	   extra1=((i-8)%10)/2;
	   k1=parts1,m1=0;			//k has to be initialized to parts for taking inner 9 seg-ends from the 10 segments v divided into. 
	   while(m1<9){
		   data1[m1][0] = frame0[k1+4+extra1][0];
		   data1[m1][1] = frame0[k1+4+extra1][1];
		   k1 += parts1;
		   m1++;
	   }
	
	   for(int l=0;l<9;l++){
		   if(l==0){
	     		a1 = data1[l][0];
	     		b1 = data1[l][1];
	   	   }
	   	   else
	     		{
	     		d1 = data1[l][0];
	     		e1 = data1[l][1];
	     		cos1= (((d1-a1)/(sqrt((d1-a1)*(d1-a1)+(e1-b1)*(e1-b1)))));
	     		sin1= (((e1-b1)/(sqrt((d1-a1)*(d1-a1)+(e1-b1)*(e1-b1)))));
	
	     		ROS_INFO(" sin:%f\tcos:%f ",sin1,cos1);
	
	     		if(cos1>0&&sin1>0)
	     			f1[l-1]=  (unsigned int) (asin(sin1)*180/PI)/10;
	     		if(cos1>0&&sin1<0)
	     			f1[l-1]=  (unsigned int) (360-180*acos(cos1)/PI)/10;
	     		if(cos1<0&&sin1>0)
	     			f1[l-1]=  (unsigned int) (180-asin(sin1)*180/PI)/10;
	     		if(cos1<0&&sin1<0)
	     			f1[l-1]=  (unsigned int) (180-asin(sin1)*180/PI)/10;
	     		a1=d1;
	     		b1=e1;
	
	     		ROS_INFO(" Angle :%f",f1[l-1]);
	     		}
	    }
	   outfile << 1 << ' ' << 16 << ' ' << 20 << "\n" ;
		ROS_INFO("Writing to file...please wait");
	      for(unsigned int p = 0; p != 8; p++){
	          outfile << f1[p] << ' ';
	      }
		for(unsigned int p = 0; p != 8; p++){
	          outfile << 70 << ' ';
	      }
	
	  	 outfile << std::endl;
	  	ROS_INFO("The Index of the Type of Gesture : ");
	  	 int n;
		scanf("%d",&n);
	  	 switch(n){
	  	 case 1 :  outfile << 1 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 2 :  outfile << 0 << ' ' << 1 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 3 :  outfile << 0 << ' ' << 0 << ' ' << 1 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 4 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 5 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 6 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 7 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 8 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 9 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 10 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 11 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 12 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 13 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 14 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 15 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 16 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 17 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 18 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 19 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 20 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << "\n" ;
				done=1;
				 break;
	
	  	 }
	}
	if(!flag[0]&&flag[1]){	 //pick for right hand
	   parts2=(j-8)/10;
	   extra2=((j-8)%10)/2;
	   k2=parts2,m2=0;
	   while(m2<9){
		   data2[m2][0] = frame1[k2+4+extra2][0];
		   data2[m2][1] = frame1[k2+4+extra2][1];
		   k2 += parts2;
		   m2++;
	   }
	
	   for(int l=0;l<9;l++){
		   if(l==0){
	     		a2 = data2[l][0];
	     		b2 = data2[l][1];
	   	   }
	   	   else
	     		{
	     		d2 = data2[l][0];
	     		e2 = data2[l][1];
	     		cos2= (((d2-a2)/(sqrt((d2-a2)*(d2-a2)+(e2-b2)*(e2-b2)))));
	     		sin2= (((e2-b2)/(sqrt((d2-a2)*(d2-a2)+(e2-b2)*(e2-b2)))));
	
	     		ROS_INFO(" sin:%f\tcos:%f ",sin2,cos2);
	
	     		if(cos2>0&&sin2>0)
	     			f2[l-1]=  (unsigned int) (asin(sin2)*180/PI)/10;
	     		if(cos2>0&&sin2<0)
	     			f2[l-1]=  (unsigned int) (360-180*acos(cos2)/PI)/10;
	     		if(cos2<0&&sin2>0)
	     			f2[l-1]=  (unsigned int) (180-asin(sin2)*180/PI)/10;
	     		if(cos2<0&&sin2<0)
	     			f2[l-1]=  (unsigned int) (180-asin(sin2)*180/PI)/10;
	     		a2=d2;
	     		b2=e2;
	
	     		ROS_INFO(" Angle :%f",f2[l-1]);
	     		}
	    }
	   outfile << 1 << ' ' << 16 << ' ' << 20 << "\n" ;
		ROS_INFO("Writing to file...please wait");
		for(unsigned int p = 0; p != 8; p++){
	          outfile << 70 << ' ';
	      }

	      for(unsigned int p = 0; p != 8; p++){
	          outfile << f2[p] << ' ';
	      }
	
	  	 outfile << std::endl;
	  	ROS_INFO("The Index of the Type of Gesture : ");
	  	 int n;
		scanf("%d",&n);
	  	 switch(n){

	  	 case 1 :  outfile << 1 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 2 :  outfile << 0 << ' ' << 1 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 3 :  outfile << 0 << ' ' << 0 << ' ' << 1 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 4 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 5 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 6 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 7 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 8 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 9 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 10 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 11 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 12 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 13 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 14 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 15 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 16 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 17 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 18 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 19 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 20 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << "\n" ;
				done=1;
				 break;

	  	 default : ROS_INFO("Wrong Value ");
	
	  	 }
	}
	if(flag[0]&&flag[1]){	 //pick for left hand and right hand
	   parts1=(i-8)/10;
	   extra1=((i-8)%10)/2;
	   k1=0,m1=0;
	   parts2=(j-8)/10;
	   extra2=((j-8)%10)/2;
	   k2=0,m2=0;
	   while(m1<9){
		   data1[m1][0] = frame0[k1+4+extra1][0];
		   data1[m1][1] = frame0[k1+4+extra1][1];
		   k1 += parts1;
		   m1++;
	   }
	 while(m2<9){
		   data2[m2][0] = frame1[k2+4+extra2][0];
		   data2[m2][1] = frame1[k2+4+extra2][1];
		   k2 += parts2;
		   m2++;
	   }
	   for(int l=0;l<9;l++){
		   if(l==0){
	     		a1 = data1[l][0];
	     		b1 = data1[l][1];
	   	   }
	   	   else
	     		{
	     		d1 = data1[l][0];
	     		e1 = data1[l][1];
	     		cos1= (((d1-a1)/(sqrt((d1-a1)*(d1-a1)+(e1-b1)*(e1-b1)))));
	     		sin1= (((e1-b1)/(sqrt((d1-a1)*(d1-a1)+(e1-b1)*(e1-b1)))));
	
	     		ROS_INFO(" sin:%f\tcos:%f ",sin1,cos1);
	
	     		if(cos1>0&&sin1>0)
	     			f1[l-1]=  (unsigned int) (asin(sin1)*180/PI)/10;
	     		if(cos1>0&&sin1<0)
	     			f1[l-1]=  (unsigned int) (360-180*acos(cos1)/PI)/10;
	     		if(cos1<0&&sin1>0)
	     			f1[l-1]=  (unsigned int) (180-asin(sin1)*180/PI)/10;
	     		if(cos1<0&&sin1<0)
	     			f1[l-1]=  (unsigned int) (180-asin(sin1)*180/PI)/10;
	     		a1=d1;
	     		b1=e1;
	
	     		ROS_INFO(" Angle :%f",f1[l-1]);
	     		}
	    }
	for(int l=0;l<9;l++){
		   if(l==0){
	     		a2 = data2[l][0];
	     		b2 = data2[l][1];
	   	   }
	   	   else
	     		{
	     		d2 = data2[l][0];
	     		e2 = data2[l][1];
	     		cos2= (((d2-a2)/(sqrt((d2-a2)*(d2-a2)+(e2-b2)*(e2-b2)))));
	     		sin2= (((e2-b2)/(sqrt((d2-a2)*(d2-a2)+(e2-b2)*(e2-b2)))));
	
	     		ROS_INFO(" sin:%f\tcos:%f ",sin2,cos2);
	
	     		if(cos2>0&&sin2>0)
	     			f2[l-1]=  (unsigned int) (asin(sin2)*180/PI)/10;
	     		if(cos2>0&&sin2<0)
	     			f2[l-1]=  (unsigned int) (360-180*acos(cos2)/PI)/10;
	     		if(cos2<0&&sin2>0)
	     			f2[l-1]=  (unsigned int) (180-asin(sin2)*180/PI)/10;
	     		if(cos2<0&&sin2<0)
	     			f2[l-1]=  (unsigned int) (180-asin(sin2)*180/PI)/10;
	     		a2=d2;
	     		b2=e2;
	
	     		ROS_INFO(" Angle :%f",f2[l-1]);
	     		}
	    }
	   outfile << 1 << ' ' << 16 << ' ' << 20 << "\n" ;
		ROS_INFO("Writing to file...please wait");
       		 for(unsigned int p = 0; p != 8; p++){
	          outfile << f1[p] << ' ';
	      }

	      for(unsigned int p = 0; p != 8; p++){
	          outfile << f2[p] << ' ';
	      }
	
	  	 outfile << std::endl;
	  	ROS_INFO("The Index of the Type of Gesture : ");
	  	 int n;
		scanf("%d",&n);
	  	 switch(n){
	  	 case 1 :  outfile << 1 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 2 :  outfile << 0 << ' ' << 1 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 3 :  outfile << 0 << ' ' << 0 << ' ' << 1 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 4 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 5 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 6 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 7 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 8 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 9 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 10 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 11 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 12 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 13 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 14 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 15 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 16 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 17 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << ' ' << 0 << "\n" ;
				done=1;
				 break;
	  	 case 18 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 19 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << ' ' << 0 << "\n" ;
	  	 	 	 done=1;
				break;
	  	 case 20 :  outfile << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' '<< 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 0 << ' ' << 1 << "\n" ;
				done=1;
				 break;
	
	  	 }
	}
 }

}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "listener");

  ros::NodeHandle n;

  ros::Publisher chatter_pub = n.advertise<gesture::gesture>("chatter", 1);
   
  ros::Subscriber sub = n.subscribe("/hand_position", 1, chatterCallback);

  ros::Rate looprate(30);

  while(!done)
  {
	ros::spinOnce();
	looprate.sleep();
  }
  return 0;
}
